src/tree/updater_quantile_hist.cc

```diff
@@ -1072,28 +1115,62 @@ inline std::pair<size_t, size_t> PartitionDenseKernel(const common::DenseColumn<
         }
       }
     }
+    }
   } else {
     for (auto rid : rid_span)  {
+    if (pgh[rid*2] != 0) {
```

Comment:
?

Reply:
It gives possibility to reduce work for partition and building histogram kernel as unused rows were marked by zero gradients. We can exclude these specific rows from build histogram process.

But currently I reverted it as the performance benefits from refactored InitSampling, UpdatePredictionCache (no exact prediction for unused rows) are much bigger than overheads for Partition and BuildLocalHistograms kernels.